### PR TITLE
Upgrade gRPC

### PR DESCRIPTION
**Pull request by vjpai**
Update commit ID of grpc, update bazel.rc to force non-use of ares resolver, and also stop patching gRPC. This should enable regular updates of gRPC since separate patches will no longer be needed.

~~Some changes below will be brought in separately by another commit and should rebase out cleanly.~~

### CODE REVIEW

**Comment by jhseu** [comment]
Will wait until the internal changes are pushed and rebase.

File: tensorflow/contrib/cmake/external/grpc.cmake [hunk 1]
```diff
@@ -28,20 +28,19 @@ else()
...
 ExternalProject_Add(grpc
     PREFIX grpc
     DEPENDS protobuf zlib
     GIT_REPOSITORY ${GRPC_URL}
     GIT_TAG ${GRPC_TAG}
     DOWNLOAD_DIR "${DOWNLOAD_LOCATION}"
     BUILD_IN_SOURCE 1
-    # TODO(jhseu): Remove this PATCH_COMMAND once grpc removes the dependency
-    # on "grpc" from the "grpc++_unsecure" rule.
-    PATCH_COMMAND ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_SOURCE_DIR}/patches/grpc/CMakeLists.txt ${GRPC_BUILD}
```
Comment (jhseu) [inline tensorflow/contrib/cmake/external/grpc.cmake]: Delete the CMakeLists.txt file in patches/grpc?

Comment (vjpai) [reply tensorflow/contrib/cmake/external/grpc.cmake]: Done.

**Comment by vjpai** [comment]
Waiting on #13976

**Comment by vjpai** [comment]
Cc: @nicolasnoble @ncteisen

**Comment by benoitsteiner** [comment]
@vjpai  #13976 is merged, but it creates a conflict with your PR. Can you fix the conflict ?

**Comment by vjpai** [comment]
Will do

**Comment by vjpai** [comment]
Ok, I think conflicts are resolved. I can haz test? (I tested it locally on bazel but not cmake, Windows, etc)

**Comment by jhseu** [comment]
Jenkins, test this please

**Review by jhseu — APPROVED** [review]
Test failures look unrelated.

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by benoitsteiner** [comment]
Jenkins, test this please

**Comment by benoitsteiner** [comment]
There are a lot of falures that look like:
tensorflow.python.framework.errors_impl.InternalError: Failed to create session.

Could that be due to the fact that this PR started to link abseil in TF, and the abseil version of the functions conflict the the ones provided by TF ?

**Comment by vjpai** [comment]
It's indeed very strange. The abseil functions and classes have their own namespace (absl::) so I don't think that there would be a build conflict; it is definitely something that needs some investigation. Right now gRPC has a dependence on abseil for future use but does not currently use any of the abseil functions.

**Comment by benoitsteiner** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by gunan** [comment]
Jenkins, test this please

**Comment by gunan** [comment]
Linux XLA failures are worrying. Just want to double check.
Jenkins, test this please.

**Comment by vjpai** [comment]
Sorry, haven't had the chance to check back in on this for a few days, but the # of failures seems high and widespread. I will try to take a look today or tomorrow. My feeling was to try to bisect the gRPC commit point or possibly push it forward since I know there have been a few bug fixes in gRPC meanwhile as well.

**Comment by gunan** [comment]
I am also surprised, I am rerunning some other tests, but XLA tests have failed on multiple runs, so I think the failures are legitimate. Did not take a hard look into why they might be though.

**Comment by vjpai** [comment]
Two speculations, both ruled out:

1. Updating the gRPC commit point will help
    - Nope! Tried this at an [updated master](https://github.com/grpc/grpc/commit/05547052d0394224515ac43364ce3a80467a5edc) and still have the same failures (which I can regenerate at my own VM using `bazel test //tensorflow/contrib/all_reduce:all_reduce_test`)
1. Removing the unneeded abseil dependence will help.
    - Nope! Made [a version of gRPC](https://github.com/vjpai/grpc/commit/7d421ae4c1be4ee160beb49dead4b6c4f20cbc48) that does this and commented out the `com_google_absl` build rule from the `workspace.bzl` but still has no effect.

**Comment by vjpai** [comment]
I've tried a few different speculations and don't believe either worked: updating the gRPC commit point and commenting out the unneeded absl dependence. Still investigating. Can you suggest a good test to consider as a signal for what's working or not? I've been trying with `bazel test //tensorflow/contrib/all_reduce:all` but I don't know how representative of a test that is.

**Comment by gunan** [comment]
I think any of the failing tests are good.
You just need to make sure you are running with XLA support enabled, on a machine that has an nvidia GPU and cuda to replicate our test setup.
Jenkins, test this please.

**Comment by gunan** [comment]
How about  `tensorflow/compiler/tests:xla_device_test`?
It seems to be a simple fast test that could help in your investigation.

**Comment by vjpai** [comment]
I speculated that maybe c-ares was being included even though it isn't supposed to be, but that's not it either. I confirmed this by putting a `#error` directive in ares specific code and confirming that I trigger it on a native gRPC build but not on a TF-specific gRPC build (since this PR defines `GRPC_ARES=0` in `tools/bazel.rc`).

**Comment by vjpai** [comment]
Ok, I now know the problem. gRPC depends on protobuf as does TF. We need to change to make gRPC depend on protobuf_headers only. Otherwise there's some sort of diamond-dependence problem which is destroying serialization. I have tested this by bringing in my own branch and confirmed that this is an issue (and hopefully the only issue) and will now get this resolved at gRPC master and then revisit this PR. Thanks for the help!

**Comment by vjpai** [comment]
The tests that were failing for me previously are now passing if I pin to my branch at grpc/grpc#13257 . I am going to push a commit now that is purely WIP because it is linked to vjpai/grpc instead of grpc/grpc , so please don't merge it. I think it will work (fingers crossed).

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by jhseu** [comment]
I'm surprised that that's an issue now. We also explicitly depend on the same version of protobuf.

**Comment by vjpai** [comment]
Hmm, looks like things are moving though! Maybe we take protobuf at a different commit point?

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by gunan** [comment]
Ubuntu contrib issue is a known flaky failure. Merging.

**Comment by gunan** [comment]
Looks like with the latest prod push, and us also adding absl dependency there is a collision.
I will try to see what is going on.

**Comment by vjpai** [comment]
For us, the absl dependence is not really important. We are not actively
using it yet. We have the dependence in place for future expansion, and
there is a strong feeling that we will actually be rolling it back and not
using it for the next quarter or so. So, feel free to change the commit
point of that dependence as you need to.

**Pull request merged**